src/lib/Type.hs

```diff
@@ -115,9 +115,31 @@ getClassTy (ClassDef _ _ bs _ _) = go bs
 
 -- === querying effects ===
 
-exprEffects :: (MonadFail1 m, EnvReader m) => Expr n -> m n (EffectRow n)
+computeAbsEffects :: (EnvExtender m, SubstE Name e)
+  => Abs (Nest Decl) e n -> m n (Abs (Nest Decl) (EffectRow `PairE` e) n)
+computeAbsEffects it = refreshAbs it \decls result -> do
+  effs <- declNestEffects decls
+  return $ Abs decls (effs `PairE` result)
+{-# INLINE computeAbsEffects #-}
+
+declNestEffects :: (EnvReader m) => Nest Decl n l -> m l (EffectRow l)
+declNestEffects decls = liftEnvReaderM $ declNestEffectsRec decls mempty
+{-# INLINE declNestEffects #-}
+
+declNestEffectsRec :: Nest Decl n l -> EffectRow l -> EnvReaderM l (EffectRow l)
+declNestEffectsRec Empty !acc = return acc
+declNestEffectsRec n@(Nest decl rest) !acc = withExtEvidence n $ do
+  expr <- sinkM $ declExpr decl
+  deff <- exprEffects expr
```

Comment:
Ohhh I wonder how much of the slowdown is actually due to `exprEffects` not getting specialized to `EnvReaderM`!

Reply:
`SPECIALIZE` pragma added.

Between the two changes, we now have
```
MUT 6.551, Total 6.974
MUT 6.614, Total 7.040
MUT 6.402, Total 6.817
MUT 6.489, Total 6.907
MUT 6.453, Total 6.876
```

which is either a significant improvement or some other process on my laptop having given up in the interim.